src/execution_plan/execution_plan.c

```diff
@@ -22,45 +22,108 @@
 }
 
-// Associate each operation in the chain with the provided RecordMap.
-static void _associateRecordMap(OpBase *root, RecordMap *record_map) {
-	// If this op has already been initialized,
-	// we don't need to recurse further.
-	if(root->record_map != NULL) return;
+static inline OpBase *_ExecutionPlan_LocateParentProjection(OpBase *root) {
+	assert(root);
+	if(root->type & (OPType_PROJECT | OPType_AGGREGATE)) return root;
+	return _ExecutionPlan_LocateParentProjection(root->parent);
+}
 
-	// Share this ExecutionPlanSegment's record map with the operation.
-	root->record_map = record_map;
+static inline OpBase *_ExecutionPlan_FindConnectingOp(OpBase *root, OpBase *prev_segment_head) {
+	// Find the leftmost leaf in this segment.
+	OpBase *leaf = _ExecutionPlan_LocateLeaf(root, prev_segment_head);
 
-	for(int i = 0; i < root->childCount; i++) {
-		_associateRecordMap(root->children[i], record_map);
+	// Traverse upwards until an aggregate/project op is found.
+	return _ExecutionPlan_LocateParentProjection(leaf);
+}
+
+static inline void _ExecutionPlan_UpdateRoot(ExecutionPlan *plan, OpBase *new_root) {
+	if(plan->root) ExecutionPlan_NewRoot(plan->root, new_root);
+	plan->root = new_root;
+}
+
+// Build the column strings for a result set.
+static inline const char **_BuildColumnNames(AR_ExpNode **exps) {
+	uint projection_count = array_len(exps);
+	const char **columns = array_new(const char *, projection_count);
+	for(uint i = 0; i < projection_count; i++) {
+		columns = array_append(columns, exps[i]->resolved_name);
+	}
+	return columns;
+}
+
+static char **_CollectAliases(rax *map) {
+	char **aliases = array_new(char *, 1);
+	raxIterator it;
+	raxStart(&it, map);
+
+	// Seek all referenced aliases prior to the anonymous identifiers.
+	raxSeek(&it, "<", (unsigned char *)"anon_", 5);
+	while(raxPrev(&it)) { // Scan backwards
+		size_t len = it.key_len;
+		// Copy the key and add a terminator character.
+		char *alias = rm_strndup((char *)it.key, len);
+		aliases = array_append(aliases, alias);
+	}
+
+	// Seek all referenced aliases after the anonymous identifiers.
+	// The ASCII value of the backtick character is 1 greater than the underscore,
+	// so this seek will skip all keys with begining with "anon_".
+	raxSeek(&it, ">", (unsigned char *)"anon`", 5);
+	while(raxNext(&it)) { // Scan forwards
+		size_t len = it.key_len;
+		// Copy the key and add a terminator character.
+		char *alias = rm_strndup((char *)it.key, len);
+		aliases = array_append(aliases, alias);
 	}
+
+	raxStop(&it);
+
+	return aliases;
 }
 
-/* In a query with a RETURN *, build projections for all explicit aliases
- * in previous clauses. */
-static void _ReturnExpandAll(ExecutionPlanSegment *segment, AST *ast) {
-	// Collect all unique aliases
-	const char **aliases = AST_CollectElementNames(ast);
+static void _PopulateProjectAll(ExecutionPlan *previous_segment, OpBase *op) {
+	char **aliases = _CollectAliases(previous_segment->record_map);
```

Comment:
`MATCH (a)-[b]->(c) WHERE a.v = 1 RETURN *`
should return `a`,`b`,`c`
although `previous_segment->record_map` would only contain `a` as it is the only identifier which is referenced.

Reply:
This query works as expected right now. The reason is that while building the AST reference map, we check to see if we're in STAR projection, and if so mark all user aliases as referenced (and must, as otherwise we won't populate AlgebraicExpressions properly).

Essentially, we handle STAR projections at the AST level and again here at the ExecutionPlan level, which is an annoying and weak approach, but will hopefully be replaced quite soon!